### PR TITLE
refactored submit project form

### PR DESCRIPTION
**Pull request by Yazdun**
### Changes proposed

The submit project page was too cluttered. I created a reusable Input component, and I completely isolated validation logic in a separate file ( single source of truth ) and pass them as prop to Input component. 

I also removed the rich text section completely, since I believe we should use markdown instead. we can save markdown as plain text in database and later, we can render markdown on the project detail page.

### Note to reviewers

> Due to Internet blackout in my country, I barely have access to internet and I wasn't able to make any API request. I implemented a fake form submission and console log the output data. Someone should take this from here and hook up the API with the form.

### CODE REVIEW

**Review by leewan-09 — CHANGES_REQUESTED** [review]
Add some margin to the bottom and top.  for the bottom, it is too less.

In my opinion, Input fields are too big in height and I don't like bold labels. 

image

File: components/FormElements/Input.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import cn from 'classnames';
+import { is_form_invalid, find_input_error } from '../../utils';
```
Comment (leewan-09) [inline components/FormElements/Input.tsx:3]: @/utils/

File: components/FormElements/Input.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  multiline,
+  validation,
+  onChange,
+}) => {
+  const {
+    register,
+    formState: { errors },
+  } = useFormContext();
+
+  const inputErrors: any = find_input_error(errors, name);
```
Comment (leewan-09) [inline components/FormElements/Input.tsx:37]: 1. Change to camelCase
2. Add type to this function

File: components/FormElements/Input.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+interface IProps {
+  name: string;
+  label: string;
+  type: string;
+  id: string;
+  placeholder: string;
+  validation?: object;
+  multiline?: boolean;
+  onChange?: any;
```
Comment (leewan-09) [inline components/FormElements/Input.tsx:16]: Add type

File: components/SubmitPageForm/SubmitForm.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
+import { FormProvider, useForm } from 'react-hook-form';
+import { Input } from '../FormElements';
+import {
+  project_name_validation,
+  project_url_validation,
+  project_desc_validation,
+  project_content_validation,
+  project_tags_validation,
+} from '../../utils/input_validators';
```
Comment (leewan-09) [inline components/SubmitPageForm/SubmitForm.tsx:9]: @/utils/

File: components/SubmitPageForm/SubmitForm.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import { useState } from 'react';
+import { AiOutlineClose } from 'react-icons/ai';
+import { InputError } from '../FormElements/Error';
+import { AnimatePresence } from 'framer-motion';
+
+export const SubmitForm = () => {
+  const methods = useForm();
+  const [tags, setTags] = useState<any>([]);
```
Comment (leewan-09) [inline components/SubmitPageForm/SubmitForm.tsx:17]: missing type

File: components/SubmitPageForm/SubmitForm.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  };
+
+  const noTag = methods.formState.isSubmitted && tags.length < 1;
+  const key = noTag ? 'keeey' : 'kjfhakjfsh';
+
+  const handleSubmit = methods.handleSubmit((data) => {
+    // successfull form submission data
+    const { tag, ...formData } = data;
+    console.log({ ...formData, tags });
```
Comment (leewan-09) [inline components/SubmitPageForm/SubmitForm.tsx:34]: remove `console.log`

File: utils/input_validators.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
+export const project_name_validation = {
```
Comment (leewan-09) [inline utils/input_validators.ts:1]: use camelCase and also follow the same file naming pattern.

**Review by leewan-09 — CHANGES_REQUESTED** [review]

File: pages/projects/submit.tsx [hunk 8]
```diff
@@ -1,12 +1,15 @@
...
+import { SubmitForm } from '@/components/SubmitPageForm/SubmitForm';
 
 const SubmitProject = () => {
   return (
     <SharedLayout title="Submit Project" hasContainer>
-      <div className="mt-4 mb-12 flex w-full flex-col space-y-4 px-0 lg:px-4">
-        <SubmitPageForm />
+      <div className="flex w-full flex-col p-4">
+        <h1 className="mb-7 border-b pb-7 text-2xl font-bold dark:border-b-gray-800">
+          Sumbit Your Project
```
Comment (rohitdasu) [inline pages/projects/submit.tsx:10]:
```suggestion
          Submit Your Project
```

**Comment by rohitdasu** [comment]
closing this PR for inactivity. Welcome to raise a PR in future 🙏

**Pull request closed without merging**